### DISCUSSION TITLE
Using dropdown in table header and table

### DISCUSSION THREAD

**Discussion by mindytom** [twilio-labs/paste#1467] [Q&A] [answered]
Hey DSYS!

I'm currently iterating on Event Stream's [Create New Subscription](https://www.figma.com/file/gWJ9De3IL9mbFW02azOQC6/2021-Segment-Sink-Type?node-id=151%3A11888) page, specifically the table. An event type (taskrouter.reservation.accepted) can support _n_ number of schema versions and I'm looking for ways to scale the current design to be more future proof. 

Right now I'm exploring ways to:
1. Select x schema version per individual event type 
(i.e. Select version 1 for taskrouter.reservation.accepted) 
2. Select x schema version and apply to all event types within the same table 
(i.e. Select version 14 for all taskrouter.reservation event types)

Feedback I'm looking for:
-  Usage of the dropdown within the table header and table
- Accessibility concerns? 
- Suggestions of using other components?

[View concept in Whimsical](https://whimsical.com/supporting-n-schemas-Ucs9f98GrnzrhLCKpTYGvB@2Ux7TurymPVqfckrMAwm)
![Table](https://user-images.githubusercontent.com/65037539/118733327-ec559d80-b7f0-11eb-80e9-38a961b16a1a.png)

**Comment by kate-porter** [comment, marked as answer]
Hey @mindytom! Thanks for sharing this for feedback. Overall, I think what you have here makes a lot of sense. The only recommendation we would make would be to pull the "Apply schema version to all" select out of the Table header and place it right above the Table instead. This will allow you to give the Schema column a static text header, which will be important for screen reader users.

One question for ya: what happens to the "Apply schema version to all" select when the user has opted to select a different version for some of the event types?

**Reply by mindytom** [reply]
Thank you @katieporter! Will make the change you suggested 👍🏼 

If a user opts to select a different version for some of the event types the "Apply schema version to all" select will go back to the default state "Select schema version". Let me know if you have a suggestion here.

**Reply by kate-porter** [reply]
Ahh okay gotcha.

Another question just for my own understanding: in the Whimsical mock you shared, there are no schema versions selected for any of the Reservation event types - if I were to click "Create subscription" in that state, is that the equivalent of not subscribing to the Reservation events at all?

**Reply by mindytom** [reply]
Yes, if a user does not have a schema version selected, that is equivalent to not subscribing to that event. 

I iterated on the wires and created this [prototype](https://www.figma.com/proto/3r5MXNo7sAUjesbVu2bfzR/Schema-versions?page-id=0%3A1&node-id=107%3A5&viewport=465%2C255%2C0.0492817685008049&scaling=min-zoom). Feel free to take a look. 

Also, it is a requirement for a user to select at least 1 event type. If a user click "Review subscription" without selecting any event types, this is what I'm thinking for the [error state](https://www.figma.com/file/3r5MXNo7sAUjesbVu2bfzR/Schema-versions?node-id=90%3A59041). Is there a recommended way to surface the error in this case?

**Reply by kate-porter** [reply]
Hey @mindytom! So sorry for the delay, I was out last week.

These prototypes are looking good. Have you had a chance to usability test these schema selection interactions yet? I'm curious to know if users find it intuitive that not choosing a schema is equivalent to not subscribing to that event. Sarah and I had played around with some ideas that split the subscription and the schema selection into two distinct actions - not sure if any of these would work at all, but wanted to share what we toyed around with: https://www.figma.com/file/SpeLpBoqN2vZv8KWcSn0J1/Feedback?node-id=406%3A33473

That error state looks good to me! Let me share this with the team to make sure we've got our bases covered from an accessibility perspective - there may be some steps you all will want to take in implementation to ensure screen reader users are alerted to that error.

**Reply by kate-porter** [reply]
Chatted with the team on the error state, and we think this looks good. Our only recommendation would be to consider adding an error toast that summarizes the errors, in addition to the inline error messaging you already have. This will help to ensure the user is aware of all errors that need to be corrected.

**Reply by mindytom** [reply]
We have done usability testing for the [radio buttons](https://www.figma.com/file/SpeLpBoqN2vZv8KWcSn0J1/Feedback?node-id=406%3A31641). We did validate that users understood selecting a schema also meant subscribing to that event type and weren't confused. 

This is what we're thinking of for the [MLP](https://www.figma.com/file/3r5MXNo7sAUjesbVu2bfzR/Schema-versions?node-id=90%3A1299). It will have only the drop downs, sticky vertical tabs, and a back to top button. We have a few customer interviews coming up so can definitely get some customer feedback on the MLP design. 

Also, love the concepts y'all explored! Opt. 3 in particular seems intriguing and something we have not tested against. Will keep it in mind as we continue to explore the future state of this flow. Thanks folks!

**Reply by mindytom** [reply]
As for the error state, is it a general recommendation to have a toast message in addition to the inline error messaging? It can feel a bit overwhelming at times if there's a ton of red on the page.

**Reply by SiTaggart** [reply]
For the error state I was personally thinking that the inline errors for the two inputs would be fine, but if you wanted to inform the user about an error for information in the table, just popping an error toast would probably be the most effective way to do so.

Inline errors are good for single fields, so it works for the inputs. But when there is an error with that gigantic table, it's not really inline anymore. It's such a large piece of UI real estate there's nothing inline about it, it's basically a page error, and showing a toast in that situation might be the most obvious and eye catching way to do it. It's no dependent on scroll position, you can be anywhere in the page and still see a contextual error, you can give it a fair amount of detail to help explain to the user what they need to do next.

**Reply by mindytom** [reply]
This is what I'm proposing for the error states for this other page [Create new sink](https://www.figma.com/file/gWJ9De3IL9mbFW02azOQC6/2021-Segment-Sink-Type?node-id=558%3A0). I decided not to show an error toast here since there won't be much of a scroll even when we have more sink type cards (~max. 2 rows) with pagination so didn't want to bombard the user with more red on the page if it wasn't needed. I'd like the error patterns to be consistent between the two pages. 

From what I understand, these are the two different recommendations I'm reading: 
1. Have inline error messages and error toast show for _any_ input, table, and card errors. 
2. Have inline error message only for inputs. Have error toasts for tables and cards. 
Am I capturing this right? @katieporter @SiTaggart. Is there a strong stance on either from DSYS? 

If that's the case for Recommendation 1, Create new sink page _can_ be a bit overwhelming with red if all inline errors and an error toast showing but this is an edge case. For Create new subscription page, having all inline errors and an error toast show can be helpful wherever the user is on the page. 
 
Recommendation 2, it may seem inconsistent to a user to see inline error messages for some errors but error toasts for others.

**Reply by SiTaggart** [reply]
The suggestion of a toast only came up due to the sheer quantity of stuff on the page and scrolling depth. Having a little inline error above a very long list of tables might get very lost.

If that's not a concern and you're actually not going to be scrolling so much, then inline errors would be fine from my perspective.

**Reply by kate-porter** [reply]
Agreed!

**Reply by mindytom** [reply]
Thank you both for your thoughtful responses. 

I'll have both the inline error messages as well as the error toast for the page with the inputs and table. As for the page with the inputs and cards I'll have only the inline error messages.
